### PR TITLE
Performance test for Config API

### PR DESCRIPTION
**Pull request by shivamkm07**
# Description

This PR adds perf test for config API.
## Issue reference

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: tests/apps/perf/configuration/app.go [hunk 1]
```diff
@@ -0,0 +1,523 @@
+/*
+Copyright 2021 The Dapr Authors
```
Comment (mukundansundar) [inline tests/apps/perf/configuration/app.go:2]: nitpick ...all new files to have latest year

File: tests/apps/perf/configuration/app.go [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+type RedisUpdater struct {
+	client *redis.Client
+}
+
+func init() {
+	p := os.Getenv("PORT")
+	if p != "" && p != "0" {
+		appPort, _ = strconv.Atoi(p)
+	}
+} // indexHandler is the handler for root path
```
Comment (mukundansundar) [inline tests/apps/perf/configuration/app.go:93]: nitpick live next line blank

File: tests/apps/perf/configuration/app.go [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+	w.WriteHeader(statusCode)
+	json.NewEncoder(w).Encode(appResponse{Message: message})
+}
+
+// return key-value pairs as a list of strings
+func getRedisValuesFromItems(items map[string]*Item) []interface{} {
+	m := make([]interface{}, 0, 2*len(items)+1)
+	for key, item := range items {
+		val := item.Value + separator + item.Version
+		m = append(m, key, val)
```
Comment (mukundansundar) [inline tests/apps/perf/configuration/app.go:112]: nitpick: Not needed in test code, but avoid append and use indexing where possible.

File: tests/apps/perf/configuration/app.go [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+	}
+	return m
+}
+
+// extract value and version from redis value
+func getRedisValueAndVersion(redisValue string) (string, string) {
```
Comment (mukundansundar) [inline tests/apps/perf/configuration/app.go:118]: same as above.

**Comment by shivamkm07** [comment]
@mukundansundar Can you run perf test on this PR to validate the changes?

**Comment by mukundansundar** [comment]
@shivamkm07 can you fix the conflicts?

**Comment by mukundansundar** [comment]
@shivamkm07 Please add the baseline results and the difference to the summary table if possible. And where are the logs written for the same?

**Comment by shivamkm07** [comment]
Latency numbers for config API:

TestConfigurationGetHTTPPerformance (229.00s)
dapr p95 latency: 81.65ms, baseline p95 latency: 19.16ms
added p95 latency: 62.49ms(326.06%)
dapr avg latency: 40.09ms, baseline avg latency: 6.91ms
added avg latency: 33.18ms(480.03%)

TestConfigurationGetGRPCPerformance (346.56s)
dapr p95 latency: 65.90ms, baseline p95 latency: 17.18ms
added p95 latency: 48.72ms(283.52%)
dapr avg latency: 31.68ms, baseline avg latency: 6.02ms
added avg latency: 25.66ms(426.28%)

TestConfigurationSubscribeHTTPPerformance (351.67s)
dapr p95 latency: 462.17ms, baseline p95 latency: 199.60ms
added p95 latency: 262.57ms(131.55%)
dapr avg latency: 236.09ms, baseline avg latency: 107.58ms
added avg latency: 128.51ms(119.46%)

TestConfigurationSubscribeGRPCPerformance (351.71s)
dapr p95 latency: 199.17ms, baseline p95 latency: 188.29ms
added p95 latency: 10.88ms(5.78%)
dapr avg latency: 106.26ms, baseline avg latency: 103.93ms
added avg latency: 2.33ms(2.24%)

File: tests/perf/configuration/configuration_test.go [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+)
+
+var (
+	tr *runner.TestRunner
+)
+
+const (
+	numHealthChecks                   = 60 // Number of times to check for endpoint health per app.
+	defaultConfigGetThresholdMs       = 500
+	defaultConfigSubscribeThresholdMs = 1000
```
Comment (mukundansundar) [inline tests/perf/configuration/configuration_test.go:41]: based on the latencies stated [here](https://github.com/dapr/dapr/pull/6247#issuecomment-1511028152), why do we have such high buffer values?

**Comment by mukundansundar** [comment]
@shivamkm07 the config perf tests seem to be failing for the second run

**Review by mukundansundar — APPROVED** [review]

**Pull request merged**